python/lib/client/dmod/client/_reader.py

```diff
@@ -0,0 +1,43 @@
+import asyncio
+from functools import partial
+from typing import Optional, Protocol, Union
+
+
+class Reader(Protocol):
+    """Reader interface type.
+
+    Example:
+    ```
+    def upload(data: Reader) -> None:
+        ...
+
+    with open("some_file.txt", "rb") as fp:
+        upload(fp)
+    ```
+    """
+
+    def read(self, size: Union[int, None] = ...) -> bytes:
```

Comment:
Any particular reason for `Union[int, None]` instead of `Optional[int]`?

Reply:
Nothing in particular, I just took inspiration from python's BufferedReader interface:

```python
def read(self, __size: int | None = ...) -> bytes: ...
```